design/LoadBalancing/LoadBalancing.md

```diff
@@ -0,0 +1,221 @@
+* Distance-based candidate election
+* Able to handle timeouts and exceptions with retries
+* etc.
+
+Currently it is used for
+
+* Storage Server interface
+* BlobWorker interface
+
+
+
+```mermaid
+graph LR
+    H((Start))
+    H0{Has alternatives?}
+    H1[Choose initial candidates]
+    H4([Never])
+    H5[Pick up an alternative]
+    H6[Send request]
+    H7[Wait for available alternative]
+    H8([Response])
+    H9([All alternatives failed])
+
+    H --> H0 --No--> H4
+    H0 --Yes--> H1
+    H1 --> H5
+    H5 --Has alternative--> H6
+    H5 --No alternative-->H7
+    H6 --Success--> H8
+    H6 --Failure--> H5
+    H7 --At least one alternative--> H5
+    H7 --> H9
+```
+
+Note:
+
+* Response could be an exception, e.g. `process_behind` or `request_maybe_delivered`, and will be delivered as `Error` to the caller.
+
+### Choose initial candidates
+
+Two initial candidates will be picked up before the requests start. They will be selected as the first two alternatives for the load balancer. If both of them failed, other alternatives are used in a round-robin way.
+
+#### No `QueueModel`
+
+If no `QueueModel` is provided, the initial candidates are picked up randomly. The first candidate, or the *best* alternative, will always be one of local workers.
+
+#### With `QueueModel`
+
+`QueueModel` holds information about each candidate related to future version, latency and penalty.
+
+* If the storage server is returning a future version error, it is marked as not available until some certain time.
+* Penalty is reported by storage server in each response (see `storageserver.actor.cpp:StorageServer::getPenalty`). It is determined by the write queue length and the version lagging.
+
+If `QueueModel` exists, the candidates will be picked base on the penalty. Workers with high penalties will be avoided when picking up the first two candidates. 
+
+### Pick up an alternative
+
+As mentioned above, the alternatives are chosen in the round-robin way when the first two candidates failed.
+
+If all alternatives failed, a flag is set, so if the next request fails with `process_behind`, the caller will receive the `process_behind` error.
+
+### Send requests to workers
+
+Here it is assumed that there are at least one alternative available.
+
+```mermaid
+graph LR
+    H((start))
+    H0{Is first request}
+    H1[Send first request]
+   	H2([Response])
+   	H3[Pick up next alternative]
+   	H4[Send additional request]
+
+    H --> H3
+    H3 -->H0
+   	H0 --Yes--> H1
+   	H1 --Success--> H2
+   	H1 --Timeout--> H3
+   	H0 --No--> H4
+   	H4 --First request succeed--> H2
+   	H4 --Second request succeed--> H2
+   	H4 --Additional request failed--> H3
+```
+
+If the first request failed, it is reset and the next request will be considered as the first request. Certain types of errors can also be returned as response, e.g. `request_may_be_delivered` or `process_behind`, which may not trigger a load-balancer retry.
```

Comment:
I believe in some situation, it sends the second request before getting the reply from the first request, is that correct?